File: 백준/Gold/20501. Facebook/Facebook.py
#Facebook
import sys
input = sys.stdin.readline

N = int(input())
links = [[] for _ in range(N)]
bitCnt = [0] * ((1 << 20) + 1)

for i in range(N):
    link = input().rstrip()

    start = 0
    end = 20
    
    while start < N:
        links[i].append(int(link[start:end], 2))
        start = end
        end += 20
    
for i in range((1 << 20) + 1):
    bitCnt[i] = i.bit_count()

Q = int(input())

for i in range(Q):
    a, b = map(int,input().split())
    cnt = 0
    
    for j in range(len(links[0])):
        cnt += bitCnt[links[a-1][j] & links[b-1][j]]

    print(cnt)
        

        

File: 백준/Gold/20501. Facebook/Facebook.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(br.readLine());
        ArrayList<ArrayList<Integer>> links = new ArrayList<>();
        int[] bitCnt = new int[(1 << 20) + 1];

        for (int i = 0; i < N; i++) {
            links.add(new ArrayList<>());
            String link = br.readLine();

            int start = 0;
            int end = 20;

            if (end > N) end = N;

            while (start < N) {
                links.get(i).add(Integer.parseInt(link.substring(start, end), 2));

                start = end;
                end += 20;

                if (end > N) end = N;
            }
        }

        for (int i = 0; i < (1 << 20) + 1; i++) {
            bitCnt[i] = Integer.bitCount(i);
        }

        int Q = Integer.parseInt(br.readLine());

        for (int i = 0; i < Q; i++) {
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int cnt = 0;

            for (int j = 0; j < links.get(0).size(); j++) {
                cnt += bitCnt[links.get(a-1).get(j) & links.get(b-1).get(j)];
            }

            sb.append(cnt).append("\n");
        }

        System.out.println(sb);
    }
}
